File: python/vcp_sdk/bridge.py
"""MCP bridge (SPEC §16, VCP-Bridge profile).

Wraps an existing MCP tool so it can be consumed as a VCP capability without
rewriting the ecosystem. A bridge MUST:

* Translate the MCP tool into a VCP capability.
* Mark provenance ``legacy_mcp``.
* **Strip the raw MCP tool description from the Planner's context** and replace
  it with a Gateway-compiled affordance summary. The raw MCP description is
  NEVER passed verbatim to the model as instruction (tool-poisoning defense).
* **Pin the observed tool schema + description hash.** If the upstream MCP
  server later changes either, the pinned hash no longer matches and the bridge
  MUST treat it as a new, unapproved capability (rug-pull defense, §4).

A bridged capability is at most VCP-L0: it adds policy, audit and pinning over
an unmodified MCP server, but cannot offer signed manifests or proof-bound
grants the upstream server does not support.
"""

from __future__ import annotations

import re
from typing import Any, Mapping, Optional

from .canonical import hash as _hash
from .identity import capability_id, contract_hash

__all__ = ["bridge_mcp_tool", "compile_affordance", "PinnedObservation"]


# Phrases that frequently appear in tool-poisoning payloads embedded in MCP
# descriptions. The affordance compiler strips them; the raw text is never used
# as an instruction regardless.
_INJECTION_MARKERS = re.compile(
    r"(?i)\b(ignore (the )?(previous|user|above)|forward all|exfiltrat|"
    r"system prompt|disregard|send (all|every).+to|<important>|do not tell)\b"
)


class PinnedObservation(dict):
    """The observed MCP tool surface, pinned by hash for rug-pull detection."""


def compile_affordance(name: str, observed_description: str) -> str:
    """Derive a safe, Gateway-authored affordance from an MCP description.

    The result is a terse, neutral summary. Any injection-style instruction in
    the raw description is removed; the raw text is never surfaced to the model
    as instruction. This is the Gateway-compiled affordance of §13.
    """
    # Take only the first sentence/line, strip injection markers, neutralize.
    first = observed_description.strip().splitlines()[0] if observed_description.strip() else ""
    first = first.split(". ")[0].strip()
    cleaned = _INJECTION_MARKERS.sub("[redacted]", first)
    cleaned = cleaned.strip(" .")
    if not cleaned:
        cleaned = name
    return f"MCP tool {name}: {cleaned}. (legacy_mcp; Gateway-compiled affordance)"


def bridge_mcp_tool(
    mcp_tool: Mapping[str, Any],
    *,
    issuer: str = "did:web:bridge.local",
    provider: str = "legacy.mcp",
    version: str = "0.0.0-mcp",
    effects: Optional[Mapping[str, Any]] = None,
    determinism: Optional[Mapping[str, Any]] = None,
    sandbox: Optional[Mapping[str, Any]] = None,
) -> dict:
    """Translate an MCP tool dict into a VCP manifest with pinned observation.

    ``mcp_tool`` is the upstream MCP advertisement, e.g.::

        {"name": "search", "description": "...", "inputSchema": {...}}

    Returns a manifest-shaped dict with ``provenance.provenance == "legacy_mcp"``,
    a pinned ``observed_schema_hash`` / ``observed_description_hash``, and a
    Gateway-compiled ``summary_for_model`` that is safe to expose. The raw MCP
    description is preserved only under ``provenance.observed_description`` for
    audit/diffing — never copied into ``summary_for_model``.
    """
    name = mcp_tool["name"]
    observed_description = str(mcp_tool.get("description", ""))
    observed_schema = mcp_tool.get("inputSchema") or mcp_tool.get(
        "input_schema", {"type": "object"}
    )

    # Pin what we observed so a later upstream mutation breaks the hash.
    observed_schema_hash = _hash(observed_schema)
    observed_description_hash = _hash(observed_description)

    # Conservative defaults: external write requiring approval unless told.
    effects = dict(
        effects
        or {
            "class": "write-irreversible",
            "external_side_effect": True,
            "requires_user_approval": True,
        }
    )
    determinism = dict(determinism or {"class": "nondeterministic"})
    sandbox = dict(sandbox or {"filesystem": "none", "network": [], "secrets": []})

    contract = {
        "issuer": issuer,
        "name": name,
        "version": version,
        "input_schema": observed_schema,
        "output_schema": {"type": "object"},
        "effects": effects,
        "determinism": determinism,
        "sandbox": sandbox,
    }
    ch = contract_hash(contract)
    cid = capability_id(contract)

    affordance = compile_affordance(name, observed_description)

    manifest = {
        "vcp": "0.1",
        "kind": "capability.manifest",
        "issuer": issuer,
        "provider": provider,
        "capability": {
            "id": cid,
            "name": name,
            "version": version,
            "contract_hash": ch,
            # User sees a neutral summary; model gets the compiled affordance.
            "summary_for_user": f"Bridged MCP tool '{name}' (legacy, unverified).",
            "summary_for_model": affordance,
            "input_schema": observed_schema,
            "output_schema": {"type": "object"},
            "effects": effects,
            "determinism": determinism,
            "sandbox": sandbox,
        },
        "provenance": {
            "provenance": "legacy_mcp",
            "source": "mcp",
            # Pinned observation for rug-pull detection (§16). Raw description is
            # kept here for diffing, NOT in summary_for_model.
            "observed_description": observed_description,
            "observed_description_hash": observed_description_hash,
            "observed_schema_hash": observed_schema_hash,
        },
        "conformance_level": "VCP-L0",
    }
    return manifest


def observation_changed(manifest: Mapping[str, Any], current_mcp_tool: Mapping[str, Any]) -> bool:
    """Return True if the upstream MCP tool drifted from the pinned observation.

    A True result means the bridge MUST treat this as a new, unapproved
    capability (rug-pull defense, §16/§4).
    """
    prov = manifest.get("provenance", {})
    cur_schema = current_mcp_tool.get("inputSchema") or current_mcp_tool.get(
        "input_schema", {"type": "object"}
    )
    cur_desc = str(current_mcp_tool.get("description", ""))
    return (
        _hash(cur_schema) != prov.get("observed_schema_hash")
        or _hash(cur_desc) != prov.get("observed_description_hash")
    )

File: rust/crates/vcp-sdk/src/bridge.rs
//! MCP bridge profile (`VCP-Bridge`, §16).
//!
//! Wraps a legacy MCP tool as a VCP capability. The two load-bearing rules:
//!
//! 1. **Never pass raw MCP text as instruction.** The upstream tool description
//!    is treated as untrusted Provider metadata. We pin its hash and synthesize
//!    a Gateway-compiled affordance for the model instead of forwarding the raw
//!    string (tool-poisoning defense, §13 / §18 test 1).
//! 2. **Pin the observed schema+description hash.** If the upstream MCP server
//!    later changes either, the pinned hash no longer matches and the bridge
//!    treats it as a new, unapproved capability (rug-pull defense, §4 / §18
//!    test 2).
//!
//! Provenance is marked `legacy_mcp`; a bridged capability is at most VCP-L0.

use serde_json::{json, Value};

use crate::identity;
use crate::jcs;
use crate::manifest::{Capability, Determinism, Effects, Manifest, Sandbox, Signature};
use crate::signer::Signer;

/// An observed MCP tool as advertised by an upstream server.
pub struct McpTool {
    pub name: String,
    /// Raw natural-language description. UNTRUSTED — never forwarded verbatim.
    pub description: String,
    pub input_schema: Value,
}

/// The result of bridging an MCP tool: a VCP manifest plus the pinned hash of
/// the observed (schema + description) so later drift is detectable.
pub struct BridgedCapability {
    pub manifest: Manifest,
    /// `sha256:` over the observed `{description, input_schema}` (the rug-pull
    /// pin). Recompute on every refresh; a mismatch means a new capability.
    pub observed_hash: String,
}

/// Compute the pinned observation hash for an MCP tool (§16: pin schema +
/// description). Bound over the raw, untrusted upstream surface so any change
/// produces a different pin.
pub fn observed_hash(tool: &McpTool) -> String {
    jcs::hash_value(&json!({
        "description": tool.description,
        "input_schema": tool.input_schema,
    }))
}

/// Bridge a legacy MCP tool into a signed VCP manifest (§16).
///
/// `provider` is the bridge's namespace (e.g. `legacy.filesystem`). The bridge
/// signs the manifest with its own key; provenance is `legacy_mcp`. The model
/// summary is Gateway-compiled and does NOT echo the raw MCP description.
pub fn bridge_mcp_tool(provider: &str, tool: &McpTool, signer: &dyn Signer) -> BridgedCapability {
    let observed = observed_hash(tool);

    // A bridged MCP tool is opaque: we cannot know its true effects, so it is
    // conservatively classed as an external side-effecting write requiring
    // approval. Sandbox is deny-all by default (the bridge adds policy + audit,
    // not real isolation, hence VCP-L0).
    let effects = Effects {
        class: "write-irreversible".to_string(),
        external_side_effect: true,
        requires_user_approval: Some(true),
        compensating_action: None,
        may_send_to: None,
        may_read_from: None,
        may_write_to: None,
    };
    let determinism = Determinism {
        class: "nondeterministic".to_string(),
        requires_idempotency_key: Some(false),
        supports_dry_run: Some(false),
    };
    let sandbox = Sandbox {
        filesystem: json!("none"),
        network: vec![],
        secrets: vec![],
    };

    let issuer = format!("bridge:{provider}");
    let contract = json!({
        "issuer": issuer,
        "name": tool.name,
        "version": "legacy",
        "input_schema": tool.input_schema,
        "output_schema": { "type": "object" },
        "effects": effects,
        "determinism": determinism,
        "sandbox": sandbox,
    });
    let contract_hash = identity::contract_hash_value(&contract);
    let cap_id = identity::capability_id(&tool.name, &contract_hash);

    // Gateway-compiled affordance: a neutral description of WHAT the capability
    // is, derived from structure, NOT the raw upstream prose. The raw text is
    // never surfaced to the Planner.
    let summary_for_model = format!(
        "Bridged legacy MCP tool '{}'. Effects are unverified (provenance legacy_mcp); \
         every write requires policy approval. Arguments must match the pinned schema.",
        tool.name
    );
    let summary_for_user = format!(
        "Legacy MCP tool '{}' (bridged, unverified). Pinned observation {}.",
        tool.name, observed
    );

    let capability = Capability {
        id: cap_id,
        name: tool.name.clone(),
        version: "legacy".to_string(),
        contract_hash,
        summary_for_user,
        summary_for_model,
        input_schema: tool.input_schema.clone(),
        output_schema: json!({ "type": "object" }),
        effects,
        determinism,
        sandbox,
        kind: Some("tool".to_string()),
    };

    let mut manifest = Manifest {
        vcp: "0.1".to_string(),
        kind: "capability.manifest".to_string(),
        issuer,
        provider: provider.to_string(),
        capability,
        provenance: Some(json!({
            "provenance": "legacy_mcp",
            "observed_hash": observed,
        })),
        signature: Signature {
            alg: signer.alg().to_string(),
            value: String::new(),
        },
    };

    let sig_value = signer.sign(manifest.signing_bytes().as_bytes());
    manifest.signature.value = sig_value;

    BridgedCapability {
        manifest,
        observed_hash: observed,
    }
}
